Sprint 9 pull request

KyleKiske opened the pull request:
Написан каркас для веб-приложения по рекомендации фильмов.
Добавлены классы User и Film и отвечающие за них контроллеры. Составлен маппинг для создания и редактирования профилей пользователей сайта и информации о фильмах.

KyleKiske committed:
Написан каркас для веб-приложения по рекомендации фильмов.
Добавлены классы User и Film и отвечающие за них контроллеры.
Составлен маппинг для создания и редактирования профилей пользователей сайта и информации о фильмах.

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,121 @@
+    private final HashSet<Film> films = new HashSet<>();
+
+    @GetMapping
+    public HashSet<Film> findAll() {
```
Тип коллекции выбирайте обощенный Map для всех реализаций HashMap, LinkedHashMap & TreeMap

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,121 @@
+        return film;
+    }
+
+    static class FilmValidationException extends Exception{
```
Эти классы - исключения лучше сделать в отдельном пакете - где находятся все исключения. Т.к. эти исключения могут понадобятся в других местах

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,121 @@
+    public Film create(@RequestBody Film film) {
+
+        for (Film filmFromSet: films){
+            if (filmFromSet.equals(film)){
```
Здесь довольно лишняя логика, т.к. если фильм есть, то он заменится просто. Просто получается что создать точно такой же фильм не получится, т.к. фильм уже есть и по этому ключу он просто заменится. 
А так мы каждый раз на ЛЮБОЙ фильм вызываем перебор коллекции фильмов - что лишние затраты ресурсов

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -0,0 +1,120 @@
+
+    private final static Logger log = LoggerFactory.getLogger(UserController.class);
+
+    private final HashSet<User> users = new HashSet<>();
```
Здесь и везде, где используете HashSet пишите так:
```
private final Map<User> users = new HashSet<>();
```

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -0,0 +1,120 @@
+    @PostMapping
+    public User create(@RequestBody User user) {
+
+        for (User userFromSet: users){
```
Здесь у меня такие же мысли как и в FilmController

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -0,0 +1,21 @@
+@Data
+@EqualsAndHashCode(onlyExplicitlyIncluded = true)
+public class Film {
+    @EqualsAndHashCode.Include
```
Здорово реализовали с ломбоком

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -0,0 +1,21 @@
+    @NotBlank
+    private String login;
+    private String name;
+    private LocalDate birthday;
```
Здесь можно поставить аннотацию валидации PastOrPresent т.к. чисто теоритически можно создать пользователя, который родился сегодня. Но не завтра

KyleKiske committed:
Исключения выделены в отдельные классы в пакете Exception.
Добавлена аннотация PastOrPresent для поля User.birthday.
Для хранения пользователей и фильмов используются Map вместо Set.

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -5,63 +5,62 @@
     private final static Logger log = LoggerFactory.getLogger(FilmController.class);
 
-    private final HashSet<Film> films = new HashSet<>();
+    private final Map<Integer, Film> films = new HashMap<>();
```
Да, верно, получается вы сейчас каждый раз по ключу получаете любой фильм со сложностью поиска 0(1)

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -5,63 +5,62 @@
-                    log.warn(exception.getMessage());
-                    return null;
-                }
+        if (films.containsKey(film.getId())){
```
На самом деле containsKey тоже содержит логику по поиску ключа, я предлагал вовсе не проверять. 
Но в целом этот вариант лучше чем был

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -70,29 +69,29 @@ public ResponseEntity<Film> putFilm(@RequestBody Film film) {
             try {
-                throw new FilmController.FilmValidationException("Дата релиза фильма раньше 28.12.1895");
-            } catch (FilmController.FilmValidationException exception){
+                throw new FilmValidationException("Дата релиза фильма раньше 28.12.1895");
```
Логи, commit-message, комментарии - лучше все писать на английском. Более универсальный подход
